src/uct/tcp/tcp_listener.c

```diff
@@ -0,0 +1,197 @@
+        status  = ucs_socket_accept(listener->listen_fd,
+                                    (struct sockaddr*)&client_addr,
+                                    &addrlen, &accept_fd);
+        if (status != UCS_OK) {
+            return;
+        }
+        ucs_assert(accept_fd != -1);
+
+        ucs_trace("server accepted a connection request from client %s",
+                  ucs_sockaddr_str((struct sockaddr*)&client_addr, ip_port_str,
+                                   UCS_SOCKADDR_STRING_LEN));
+
+        /* Set the accept_fd to non-blocking mode
+         * (so that send/recv won't be blocking) */
+        status = ucs_sys_fcntl_modfl(accept_fd, O_NONBLOCK, 0);
+        if (status != UCS_OK) {
+            goto err;
+        }
+
+        sa_arg_ctx = ucs_malloc(sizeof(uct_tcp_sa_arg_t), "accept sa_arg_ctx");
+        if (sa_arg_ctx == NULL) {
+            ucs_error("failed to allocate memory for a listen_ctx");
+            goto err;
+        }
+
+        sa_arg_ctx->fd = accept_fd;
+        async_ctx      = listener->super.cm->iface.worker->async;
+
+        UCS_ASYNC_BLOCK(async_ctx);
+        ucs_list_add_tail(&listener->sockcm->sa_arg_list, &sa_arg_ctx->list);
+        UCS_ASYNC_UNBLOCK(async_ctx);
+
+        status = ucs_async_set_event_handler(async_ctx->mode, accept_fd,
+                                             UCS_EVENT_SET_EVREAD |
+                                             UCS_EVENT_SET_EVERR,
+                                             uct_tcp_sa_data_handler,
+                                             sa_arg_ctx, async_ctx);
+        if (status != UCS_OK) {
+            goto err_free_ctx;
+        }
+    }
+
+err_free_ctx:
+    ucs_free(sa_arg_ctx);
+err:
+    close(accept_fd);
+}
+
+UCS_CLASS_INIT_FUNC(uct_tcp_listener_t, uct_cm_h cm,
+                    const struct sockaddr *saddr, socklen_t socklen,
+                    const uct_listener_params_t *params)
+{
+    uct_tcp_sockcm_t *tcp_sockcm  = ucs_derived_of(cm, uct_tcp_sockcm_t);
+    ucs_async_context_t *async_ctx;
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+    int backlog;
+
+    UCS_CLASS_CALL_SUPER_INIT(uct_listener_t, cm);
+
+    self->sockcm          = tcp_sockcm;
+    self->conn_request_cb = params->conn_request_cb;
+    self->user_data       = (params->field_mask & UCT_LISTENER_PARAM_FIELD_USER_DATA) ?
+                            params->user_data : NULL;
+    backlog               = (params->field_mask & UCT_LISTENER_PARAM_FIELD_BACKLOG) ?
+                            params->backlog : ucs_socket_max_conn();
+
+    status = ucs_socket_create(saddr->sa_family, SOCK_STREAM, &self->listen_fd);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    /* Set the listener_fd to non-blocking mode
+     * (so that accept() won't be blocking) */
+    status = ucs_sys_fcntl_modfl(self->listen_fd, O_NONBLOCK, 0);
+    if (status != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
```

Comment:
looks like this was copy-pasted by mistake
don't need this, since `ucs_sys_fcntl_modfl()` returns `UCS_ERR_IO_ERROR` in case of error